test/benchmark/l4bench.go

```diff
@@ -0,0 +1,194 @@
+package benchmarktest
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/cilium/cilium/pkg/logging"
+	. "github.com/cilium/cilium/test/ginkgo-ext"
+	"github.com/cilium/cilium/test/helpers"
+
+	. "github.com/onsi/gomega"
+	"github.com/sirupsen/logrus"
+)
+
+var _ = Describe("BenchmarkNetperfPerformance", func() {
+	var (
+		vm          *helpers.SSHMeta
+		monitorStop = func() error { return nil }
+
+		log    = logging.DefaultLogger
+		logger = logrus.NewEntry(log)
+
+		PerfLogFile   = "l4bench_perf.log"
+		PerfLogWriter bytes.Buffer
+	)
+
+	BeforeAll(func() {
+		vm = helpers.InitRuntimeHelper(helpers.Runtime, logger)
+		helpers.ExpectCiliumReady(vm)
+	})
+
+	AfterAll(func() {
+	})
+
+	JustBeforeEach(func() {
+		monitorStop = vm.MonitorStart()
+	})
+
+	JustAfterEach(func() {
+		vm.ValidateNoErrorsInLogs(CurrentGinkgoTestDescription().Duration)
+		Expect(monitorStop()).To(BeNil(), "cannot stop monitor command")
+	})
+
+	AfterFailed(func() {
+		vm.ReportFailed(
+			"cilium service list",
+			"cilium policy get")
+	})
+
+	AfterEach(func() {
+		LogPerm := os.FileMode(0666)
+		testPath, err := helpers.CreateReportDirectory()
+		Expect(err).Should(BeNil(), "cannot create log file")
+		helpers.WriteOrAppendToFile(filepath.Join(testPath, PerfLogFile), PerfLogWriter.Bytes(), LogPerm)
+		PerfLogWriter.Reset()
+	}, 500)
+
+	createContainers := func() {
+		vm.ContainerCreate(helpers.Client, helpers.NetperfImage, helpers.CiliumDockerNetwork, "-l id.client")
+		vm.ContainerCreate(helpers.Server, helpers.NetperfImage, helpers.CiliumDockerNetwork, "-l id.server")
+		vm.PolicyDelAll()
+		vm.WaitEndpointsReady()
+		err := helpers.WithTimeout(func() bool {
+			if data, _ := vm.GetEndpointsNames(); len(data) < 2 {
+				logger.Info("Waiting for endpoints to be ready")
+				return false
+			}
+			return true
+		}, "Endpoints are not ready", &helpers.TimeoutConfig{Timeout: 150})
+		Expect(err).Should(BeNil())
+	}
+
+	removeContainers := func(containerName string) {
+		By(fmt.Sprintf("removing container %s", containerName))
+		res := vm.ContainerRm(containerName)
+		Expect(res.WasSuccessful()).Should(BeTrue())
+	}
+
+	deleteContainers := func() {
+		removeContainers(helpers.Client)
+		removeContainers(helpers.Server)
+	}
+
+	BeforeEach(func() {
+	}, 500)
+
+	Context("Benchmark Netperf Tests Container2Container", func() {
+		BeforeAll(func() {
+			createContainers()
+		})
+
+		AfterAll(func() {
+			deleteContainers()
+		})
+
+		ClientServerCheck := func(serverIPv6 string, serverIP string) {
+			By("checking %s can ping to %s IPv6", helpers.Client, helpers.Server)
+			res := vm.ContainerExec(helpers.Client, helpers.Ping6(serverIPv6))
+			res.ExpectSuccess()
```

Comment:
When this `ExpectSuccess()` fails, the failure in the ginkgo output will print that this line is the point of failure and it might not be obvious where the failure actually occurred from the test context (eg the fact that it was the 5th `SuperNetperfRR` call).

We will often work around this by doing something like:

```
ExpectWithOffset(1, res).To(helpers.CMDSuccess(), "Connectivity check failed")
```

Where 1 is the distance from this function to the actual test function (`Describe()`, `It()`, ...) that is calling this helper.

Same feedback applies to the rest of this helper and the other helper functions you've defined in this Context.

Reply:
Now that everything is broken into It() branches and netperf code pulled into helpers/benchmark I think we can skip ExpectWithOffset and it will still be clear what failed and where.